--- eggs/Deliverance-0.3c4-py2.6.egg/deliverance/proxy.py	2010-06-10 15:02:17.000000000 +0300
+++ eggs/Deliverance-0.3c5dev-py2.6.egg/deliverance/proxy.py	2010-06-10 15:00:33.000000000 +0300
@@ -620,11 +620,11 @@
             def link_repl_func(link):
                 """Rewrites a link to point to this proxy"""
                 if link == exact_proxied_base:
-                    return exact_orig_base
+                    pass
                 if not link.startswith(proxied_base):
                     # External link, so we don't rewrite it
                     return link
-                new = orig_base + link[len(proxied_base):]
+                new = "/" + link[len(proxied_base):]
                 return new
             if response.content_type != 'text/html':
                 log.debug(
@@ -637,10 +637,19 @@
                     body = response.body
                 else:
                     body = response.unicode_body
-                body_doc = document_fromstring(body, base_url=proxied_url)
-                body_doc.make_links_absolute()
-                body_doc.rewrite_links(link_repl_func)
-                response.body = tostring(body_doc)
+                try:
+                    body_doc = document_fromstring(body, base_url=proxied_url)
+                except:
+                    try:
+                        body_doc = document_fromstring(body.encode(response.charset), base_url=proxied_url)
+                    except:
+                        body_doc = None
+                if body_doc:
+                    body_doc.make_links_absolute()
+                    body_doc.rewrite_links(link_repl_func)
+                    response.body = tostring(body_doc)
+                else:
+                    response.body = ""
             if response.location:
                 ## FIXME: if you give a proxy like
                 ## http://openplans.org, and it redirects to
@@ -649,9 +658,10 @@
                 ## rewritten, but some better log message is required
                 loc = urlparse.urljoin(proxied_url, response.location)
                 loc = link_repl_func(loc)
-                response.location = loc
+                ## FIXME: How to rewrite response.locations to dv.xdvserver's port?
+                response.location = "http://localhost:8000" + loc
             if 'set-cookie' in response.headers:
-                cookies = response.headers.getall('set-cookie')
+                cookies = [cookie.replace("; Secure", ";") for cookie in response.headers.getall('set-cookie')]
                 del response.headers['set-cookie']
                 for cook in cookies:
                     old_domain = urlparse.urlsplit(proxied_url)[1].lower()
